ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp

```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
+    // move the kernel horizontally one stride to the right
+    kernel_position[1] += kernel_strides[1];
+
+    // if the top-right corner of the kernel is outside of the padding area,
+    // move it back to the left and one stride down
+    if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+        kernel_position[1] = 0 - pads_begin[1];
+        kernel_position[0] += kernel_strides[0];
+    }
+
+    return kernel_position;
+}
+
+template <typename T>
+Coord<T> next_kernel_position_3D(Coord<T> kernel_position,
+                                 const Shape& kernel,
+                                 const Strides& kernel_strides,
+                                 const Strides& kernel_dilations,
+                                 const Shape& data_shape,
+                                 const Shape& pads_begin,
+                                 const Shape& pads_end) {
+    kernel_position[2] += kernel_strides[2];
+
+    if (kernel_position[2] + (kernel[2] - 1) * kernel_dilations[2] >= data_shape[4] + pads_end[2]) {
+        kernel_position[2] = 0 - pads_begin[2];
+        kernel_position[1] += kernel_strides[1];
+        if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+            kernel_position[1] = 0 - pads_begin[1];
+            kernel_position[0] += kernel_strides[0];
+        }
+    }
+
+    return kernel_position;
+}
+
+namespace kernel {
+template <typename Values_t, typename Indices_t>
+void max_pool_1d(const Values_t* data,
+                 Values_t* values,
+                 Indices_t* indices,
+                 const size_t data_elems,
+                 const size_t out_elems,
+                 const size_t kernel_size,
+                 const size_t kernel_stride,
+                 const size_t kernel_dilation,
+                 const size_t pads_begin,
+                 const size_t pads_end,
+                 const size_t indices_offset) {
+    int kernel_position = 0 - pads_begin;
+    // select max elem and its index for each "placeholder" in the out buffer (pointed to by out_idx)
+    for (size_t out_idx = 0; out_idx < out_elems; ++out_idx) {
+        Values_t max_elem = std::numeric_limits<Values_t>::lowest();
+        Indices_t max_elem_idx = Indices_t{0};
+        for (size_t kernel_elem = 0; kernel_elem < kernel_size; ++kernel_elem) {
+            const size_t kernel_elem_offset = kernel_elem * kernel_dilation;
+            if (kernel_position + kernel_elem_offset < 0 || kernel_position + kernel_elem_offset >= data_elems) {
+                // don't process the padding elements
+                continue;
+            } else {
+                if (data[kernel_position + kernel_elem_offset] > max_elem) {
+                    max_elem = data[kernel_position + kernel_elem_offset];
+                    max_elem_idx = kernel_position + kernel_elem_offset;
+                }
+            }
```

Comment:
The branch of `if` statement which will likely be true should not be in `else` branch. This can result in severe performance degradation due to branch prediction in the CPU. So, I would insist on refactoring this code to:

```cpp
            if (kernel_position + kernel_elem_offset >= 0 && kernel_position + kernel_elem_offset < data_elems && data[kernel_position + kernel_elem_offset] > max_elem) {
                max_elem = data[kernel_position + kernel_elem_offset];
                max_elem_idx = kernel_position + kernel_elem_offset;
            }
```